### PR TITLE
Snapshots documentation

### LINKED ISSUE #14: Back up databases to S3-compatible storage
There are many options for backing up SQLite to S3. The one that most directly focuses on continuous streaming backups and restore (as opposed to maintaining read replicas or multiple writers, etc.) is [litestream](https://litestream.io/getting-started/), but there are others like [libSQL's bottomless replication work-in-progress](https://github.com/tursodatabase/libsql/tree/main/bottomless).

We discuss alternatives and why we aren't (yet) considering them below, but for now, we will go with [simple scheduled backups](https://litestream.io/alternatives/cron/).
* In `ayb.toml`, you will configure: S3-compatible storage details, frequency of backup, and how many backups to keep before pruning old ones.
* `ayb` will have a Snapshot data model that can store details on database snapshots (when we took it, what the hash of the database directory was, whether it was a scheduled or manual snapshot, the path to the backup on S3-compatible storage, etc.)
* On the `ayb` server, we'll run a periodic task to check if a database has changed recently (based on the hash of its database directory, but perhaps there's a way to query for the most recent transaction ID instead) and if it has, create a new snapshot.
* The backup in SQLite will happen via [VACUUM](https://www.sqlite.org/lang_vacuum.html), which will reduce file size and won't potentially run forever under high write loads like the [online backup API](https://www.sqlite.org/backup.html).
* The advantages of this approach are:
  * Simplest implementation path.
  * No dependence on other tools that are being rapidly developed and don't have documented APIs (yet).
  * Due to the simplicity and generality of the approach, it shouldn't be hard to apply it to other database types (e.g., DuckDB) that have less developed stream-the-WAL-to-S3 functionality.
  * I suspect that in practice, most databases will be write-dormant most of the time given the target use cases, so this might require less overhead than tools that constantly monitor a database log for changes.
* The disadvantages of the approach are:
  * Potentially more work per backup: if you only update a row, we still VACUUM and export the entire database.
  * Periodic snapshots have a larger window for data loss on the order of tens of minutes to an hour, whereas shipping the write-ahead log typically results in only seconds of data loss.
  * Related to the previous reason, implementing things like multi-node `ayb` support with automatic failover will be difficult in the future in cases where a node becomes inaccessible and you're not sure how many minutes/tens of minutes of backups are unsynced.

The plan
- [x] Ask on SQLite forum about how to find the most recent txid/modification: https://sqlite.org/forum/forumpost/2798df4be8. The approach to beat is to have each DB in its own directory (SQLite spans multiple files), and use filesystem modified times / hashes of all files in the directory to determine changes.
- [x] Refactors to directory structure that support snapshots and SQLite-in-directory (https://github.com/marcua/ayb/pull/311)
- [x] Refactors to allow dangerous/un-isolated requests to complete the backups (requires ATTACH) (https://github.com/marcua/ayb/pull/312)
- [x] Backups to the local filesystem (https://github.com/marcua/ayb/pull/316)
  - [x] `periodic_snapshots` configuration option with details like snapshot frequency and quantity before pruning, as well as S3-compatible connection information
  - [x] Daemon/worker thread to periodically run backups using VACUUM approach
- [x] S3 management to list/create snapshots (https://github.com/marcua/ayb/pull/317)
  - [x] Snapshot model with: primary key, database foreign key, creation date, snapshot ID (database hash/transaction ID, indexed), automatic/manual enum, [future not now: nullable note field for manual snapshot]
- [x] end-to-end tests using localstack as stand-in for S3 (https://github.com/marcua/ayb/pull/358)
- [x] Endpoints / client support for listing snapshots (#385)
- [x] Endpoints / client support for restoring from snapshots (#405)
- [x] Hash snapshot content to avoid uploading already-snapshotted database (#417 and #418)
- [x] Remove old snapshots in list beyond the retention quantity/period. (#418)
- [x] Replace gzip with zstd (#424)
- [x] Clean up: Initialize a HostedDb that has a SQLite / DuckDB implementation. Push query/backup logic into that. Consider doing this on an InstantiatedDatabase directly. (Turned into #426) 
- [x] Optimization based on modified time to reduce snapshots: get fs::metadata of each file in the dir, call `modified()` on result, sort the times so it's stable, take max or hash them together. (Turned into #427)
- [x] Final cleanups (#432)
  - [x] Clean up TODOs, especially in src/server/snapshots.rs
  - [x] Can we replace walkdir with std::fs::read_dir
- [x] Add a mutex to prevent overlapping/concurrent snapshot runs (#444)
- [ ] Write docs

## Alternatives we considered
* Knowing nothing,  [litestream](https://litestream.io/getting-started/) seemed like the most obvious option. The litestream documentation isn't clear on how many databases worth of replication it's designed for, but given that adding a database requires a configuration file update and a litestream restart, it was more likely designed around single digit to tens of databases. I roughly confirmed this in [a discussion in the libstream board](https://github.com/benbjohnson/litestream/discussions/561). The discussion raised the idea of waiting on a third party to open source one of their services for solving a similar problem, or calling Go from Rust, which we can consider in the future as those projects mature.
* In researching this work, the [libSQL](https://github.com/tursodatabase/libsql) project, and in particular 
[libsql-server/sqld](https://github.com/tursodatabase/libsql/tree/main/libsql-server) and its [bottomless replication support](https://github.com/tursodatabase/libsql/tree/main/bottomless) look promising. The bottomless documentation and commit history suggests its under quite a bit of heavy development. After all of my research, my intuition is that if the team ships a stable `bottomless` library, we'll be able to use it directly from `ayb` (both are in Rust) to solve the stream-the-write-ahead-log-to-and-recover-from-S3 problem.

Both projects, but in particular the second one, seem promising for the future, but weren't directly available for import at the time of writing, so for this v1 of backup/restore, we'll fall back to a [simpler solution like scheduled backups](https://litestream.io/alternatives/cron/).

### PR DESCRIPTION
**Pull request by marcua**
This PR resolves #14 (!!!!!!!!) by adding documentation

### CODE REVIEW

**Comment by sofiaritz** [comment]
Yayy!!! Contrata :)

**Comment by marcua** [comment]
Thank you @sofiaritz! You are too kind! :)

File: README.md [hunk 1]
```diff
@@ -166,6 +166,91 @@ $ curl -w "\n" -X POST http://127.0.0.1:5433/v1/marcua/test.sqlite/query -H "aut
 {"fields":["name","score"],"rows":[["PostgreSQL","10"],["SQLite","9"],["DuckDB","9"]]}
 ```
 
+### Snapshots / backups
+
+You can configure `ayb` to periodically upload snapshots of each
+database to [S3](https://aws.amazon.com/s3/)-compatible storage to
+recover from the failure of the machine running `ayb` or revert to a
+previous copy of the data. Each snapshot is compressed (using
+[zstd](https://zstd.net)) and only uploaded if the database changed
```
Comment (marcua) [inline README.md:175]:
```suggestion
[zstd](https://facebook.github.io/zstd/)) and only uploaded if the database changed
```

File: README.md [hunk 2]
```diff
@@ -166,6 +166,91 @@ $ curl -w "\n" -X POST http://127.0.0.1:5433/v1/marcua/test.sqlite/query -H "aut
...
+[snapshots.automation]
+interval = "10m"
+max_snapshots = 3
+```
+
+Here is an explanation of the parameters:
+* `sqlite_method`: The two SQLite backup methods are [Vacuum](https://www.sqlite.org/lang_vacuum.html#vacuuminto) and [Backup](https://www.sqlite.org/backup.html). `ayb` only supports `Vacuum` for now.
+* `access_key_id` / `secret_access_key`: The access key ID and secret to upload/list snapshots to your S3-compatible storage provider.
+* `bucket`: The name of the bucket to which to upload snapshots.
+* `bucket_prefix`: (Can be blank) if you want to upload snapshots to a prefixed path inside `bucket` (e.g., `my-bucket/snapshots`), provide a prefix.
```
Comment (marcua) [inline README.md:199]:
```suggestion
* `bucket_prefix`: (Can be blank) if you want to upload snapshots to a prefixed path inside `bucket` (e.g., `my-bucket/the-snapshots`), provide a prefix (e.g., `the-snapshots`).
```

File: README.md [hunk 3]
```diff
@@ -166,6 +166,91 @@ $ curl -w "\n" -X POST http://127.0.0.1:5433/v1/marcua/test.sqlite/query -H "aut
...
+* `access_key_id` / `secret_access_key`: The access key ID and secret to upload/list snapshots to your S3-compatible storage provider.
+* `bucket`: The name of the bucket to which to upload snapshots.
+* `bucket_prefix`: (Can be blank) if you want to upload snapshots to a prefixed path inside `bucket` (e.g., `my-bucket/the-snapshots`), provide a prefix (e.g., `the-snapshots`).
+* `endpoint_url`: (Optional if using AWS S3) Each S3-compatible storage provider will tell you their own endpoint to manage your buckets.
+* `region`: (Optional if using AWS S3) Some S3-compatible storage providers will request a region in their network where your bucket will live.
+* `force_path_style`: (Optional, legacy) If included and `true`, will use the legacy [path-style](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html#path-style-access) method of referencing buckets. Used in `ayb`'s end-to-end tests and might be helpful beyond, but start without it.
+* `interval`: How frequently to take a snapshot of your data in human-readable format (e.g., `30m`, `1h`, `1h30m`, with [more examples here](https://docs.rs/go-parse-duration/latest/go_parse_duration/)).
```
Comment (marcua) [inline README.md:203]:
```suggestion
* `interval`: How frequently to take a snapshot of your data in human-readable format (e.g., every 30 minutes = `30m`, every hour = `1h`, every hour and 30 minutes = `1h30m`, with [more examples here](https://docs.rs/go-parse-duration/latest/go_parse_duration/)).
```

File: README.md [hunk 4]
```diff
@@ -166,6 +166,91 @@ $ curl -w "\n" -X POST http://127.0.0.1:5433/v1/marcua/test.sqlite/query -H "aut
...
+Rows: 2
+
+# Wait longer than 3 seconds before listing snapshots to ensure that a snapshot with SQLite exists as well.
+$ ayb client list_snapshots marcua/snapshots.sqlite
+ Name                                                             | Last modified
+------------------------------------------------------------------+---------------------------
+ f9e01a396fb7f91be988c26d43f9ffa667bd0fd05009b231aa61ea1073d34423 | 2024-08-18T15:05:04+00:00
+ 856e21f7cae8383426cd2e0599caf6e83962b051af4734ab5c53aff87ea0ff45 | 2024-08-18T15:04:40+00:00
+
+# Restire the older snapshot, which didn't contain SQLite
```
Comment (marcua) [inline README.md:237]:
```suggestion
# Restore the older snapshot, which didn't contain SQLite
```

**Pull request merged**